### PR TITLE
Store changes to persist data columns

### PR DESCRIPTION
**Pull request by jimmygchen**
## Issue Addressed

Part of #6072.

This PR contains only the Store changes to persist data columns.

DataColumns are stored individually by index. Sampling peers will request single columns or arbitrary slots. Storing bundled like blobs would be inefficient. 

Upstream PRs:
- https://github.com/sigp/lighthouse/pull/5050
- https://github.com/sigp/lighthouse/pull/5890

### CODE REVIEW

File: beacon_node/store/src/lib.rs [hunk 1]
```diff
@@ -143,6 +142,28 @@ pub fn get_key_for_col(column: &str, key: &[u8]) -> Vec<u8> {
...
 }
 
+pub fn get_data_column_key(block_root: &Hash256, column_index: &ColumnIndex) -> Vec<u8> {
+    let mut result = block_root.as_bytes().to_vec();
+    result.extend_from_slice(&column_index.to_le_bytes());
+    result
+}
+
+pub fn parse_data_column_key(data: Vec<u8>) -> Result<(Hash256, ColumnIndex), Error> {
+    if data.len() != 32 + 8 {
```
Comment (realbigsean) [inline beacon_node/store/src/lib.rs]: we can use `DBColumn::BeaconDataColumn.key_size()` here

File: beacon_node/store/src/hot_cold_store.rs [hunk 2]
```diff
@@ -1748,6 +1914,24 @@ impl<E: EthSpec, Hot: ItemStore<E>, Cold: ItemStore<E>> HotColdDB<E, Hot, Cold>
         self.blob_info.read_recursive().clone()
     }
 
+    /// Initialize the `DataColumnInfo` when starting from genesis or a checkpoint.
+    pub fn init_data_column_info(&self, anchor_slot: Slot) -> Result<KeyValueStoreOp, Error> {
+        let oldest_data_column_slot = self.spec.deneb_fork_epoch.map(|fork_epoch| {
```
Comment (realbigsean) [inline beacon_node/store/src/hot_cold_store.rs]: should this be pectra fork epoch

Comment (jimmygchen) [reply beacon_node/store/src/hot_cold_store.rs]: Good point - currently we're testing peerdas on top of deneb, so we've been using deneb in a few places, but this should be `eip7594_fork_epoch` and we can change it to electra later once we rebase PeerDAS on `electra`.

Comment (jimmygchen) [reply beacon_node/store/src/hot_cold_store.rs]: Fixed in https://github.com/sigp/lighthouse/pull/6073/commits/9b2016875a3922331e5390e9ed7f115df87479b3

File: beacon_node/store/src/lib.rs (hunk 1 above)
Comment (jimmygchen) [reply beacon_node/store/src/lib.rs]: ah nice, thanks! Fixed in https://github.com/sigp/lighthouse/pull/6073/commits/9b2016875a3922331e5390e9ed7f115df87479b3

File: beacon_node/store/src/hot_cold_store.rs [hunk 3]
```diff
@@ -294,11 +317,39 @@ impl<E: EthSpec> HotColdDB<E, LevelDB<E>, LevelDB<E>> {
...
+                let oldest_data_column_slot = data_column_info
+                    .oldest_data_column_slot
+                    .or(eip7594_fork_slot);
+                DataColumnInfo {
+                    oldest_data_column_slot,
+                }
+            }
+            // First start.
+            None => DataColumnInfo {
+                // Set the oldest data column slot to the Deneb fork slot if it is not yet set.
```
Comment (michaelsproul) [inline beacon_node/store/src/hot_cold_store.rs]:
```suggestion
                // Set the oldest data column slot to the fork slot if it is not yet set.
```

File: beacon_node/store/src/hot_cold_store.rs [hunk 4]
```diff
@@ -294,11 +317,39 @@ impl<E: EthSpec> HotColdDB<E, LevelDB<E>, LevelDB<E>> {
...
         db.compare_and_set_blob_info_with_write(<_>::default(), new_blob_info.clone())?;
+
+        let data_column_info = db.load_data_column_info()?;
+        let eip7594_fork_slot = db
+            .spec
+            .eip7594_fork_epoch
+            .map(|epoch| epoch.start_slot(E::slots_per_epoch()));
+        let new_data_column_info = match &data_column_info {
+            Some(data_column_info) => {
+                // Set the oldest data column slot to the Deneb fork slot if it is not yet set.
```
Comment (michaelsproul) [inline beacon_node/store/src/hot_cold_store.rs]:
```suggestion
                // Set the oldest data column slot to the fork slot if it is not yet set.
```

**Review by michaelsproul — APPROVED** [review]
LGTM!

Although this introduces new columns on disk I think it is OK to merge this without a schema version increment as nothing will be stored in those columns until the fork epoch is set.

Ideally we would want to block downgrading to a column-oblivious schema once peer DAS is activated, but it's kind of a moot point because nodes won't be able to follow the chain without these changes (and the rest of the DAS logic).

We will also be bumping the schema version (and disallowing downgrades) with this PR:

- https://github.com/sigp/lighthouse/pull/5978

**Comment by michaelsproul** [comment]
@mergify queue

**Comment by michaelsproul** [comment]
@mergify refresh

**Comment by michaelsproul** [comment]
@mergify queue

**Pull request merged**